File: scorecard/core/calculator.py
import json
import os

class BVSSCalculator:
    def __init__(self, config_path=None):
        if config_path is None:
            # Default path relative to this file
            config_path = os.path.join(os.path.dirname(__file__), '..', 'scorecard_config.json')
        
        with open(config_path, 'r', encoding='utf-8') as f:
            self.config = json.load(f)
        
        self.logic = self.config['scoring_logic']
        self.factors = self.config['blockchain_factors']

    def calculate(self, impact, likelihood, exploitability='Network', scope='Unchanged', economic_loss='Millions', is_immutable=True, exploit_maturity='POC', privileged_access='Not Required'):
        # 1. BVSS Base Score: Average of core metrics (Impact, Likelihood, Exploitability, Economic Loss)
        impact_val = self.logic['impact_map'].get(impact, 1)
        likelihood_val = self.logic['likelihood_map'].get(likelihood, 1)
        expl_val = self.logic['bvss_metrics']['exploitability_map'].get(exploitability, 1.0)
        econ_val = self.logic['bvss_metrics']['economic_loss_map'].get(economic_loss, 1)
        scope_mult = self.logic['bvss_metrics']['scope_map'].get(scope, 1.0)
        
        base_score = (impact_val * 0.25 + likelihood_val * 0.25 + expl_val * 0.25 + econ_val * 0.25) * scope_mult
        
        # 2. Apply Immutability Multiplier
        if is_immutable:
            base_score *= self.logic['multipliers']['immutability']
            
        # 3. Apply Blockchain Factors
        base_score *= self.factors['exploit_maturity'].get(exploit_maturity, 1.0)
        base_score *= self.factors['privileged_access'].get(privileged_access, 1.0)
        
        # 4. Dynamic Scale to 0-10 (BVSS max possible)
        max_impact = 4
        max_lik = 3
        max_expl = max(self.logic['bvss_metrics']['exploitability_map'].values())
        max_econ = max(self.logic['bvss_metrics']['economic_loss_map'].values())
        max_base = (max_impact + max_lik + max_expl + max_econ) / 4.0
        max_scope = max(self.logic['bvss_metrics']['scope_map'].values())
        max_imm = self.logic['multipliers']['immutability']
        max_mat = max(self.factors['exploit_maturity'].values())
        max_priv = max(self.factors['privileged_access'].values())
        max_possible = max_base * max_scope * max_imm * max_mat * max_priv
        
        scaled_score = (base_score / max_possible) * 10
        scaled_score = min(max(round(scaled_score, 1), 0), 10)
        
        # 5. Determine Severity
        severity = "Low"
        color = "🟢"
        for level in self.logic['severity_levels']:
            if scaled_score >= level['threshold']:
                severity = level['label']
                color = level['color']
                break
                
        return {
            "score": scaled_score,
            "severity": severity,
            "color": color
        }

if __name__ == "__main__":
    calc = BVSSCalculator()
    print(calc.calculate("Critical", "High", True, "Active", "Not Required"))

File: apps/web/lib/calculator.ts
import { BVSSParams, ScoreResult } from './types';

export class BVSSCalculator {
  private config: any;
  private logic: any;
  private factors: any;

  constructor(config: any) {
    this.config = config;
    this.logic = config.scoring_logic;
    this.factors = config.blockchain_factors;
  }

  public calculate(params: Partial<BVSSParams>): ScoreResult {
    // Apply defaults for fields that might be missing
    const impact = params.impact || 'Medium';
    const lik = params.likelihood || 'Medium';
    const expl = params.exploitability || 'Network';
    const scopeVal = params.scope || 'Unchanged';
    const econ = params.economicLoss || 'Millions';
    const isImm = params.isImmutable !== undefined ? params.isImmutable : true;
    const maturity = params.exploitMaturity || 'POC';
    const priv = params.privilegedAccess || 'Not Required';
    const autonomy = params.agentAutonomy || 'None';

    // 1. BVSS Base Score: Average of core metrics
    const impactVal = this.logic.impact_map[impact] || 1;
    const likelihoodVal = this.logic.likelihood_map[lik] || 1;
    const explVal = this.logic.bvss_metrics.exploitability_map[expl] || 1.0;
    const econVal = this.logic.bvss_metrics.economic_loss_map[econ] || 1;
    const scopeMult = this.logic.bvss_metrics.scope_map[scopeVal] || 1.0;
    
    let baseScore = (impactVal * 0.25 + likelihoodVal * 0.25 + explVal * 0.25 + econVal * 0.25) * scopeMult;
    
    // 2. Apply Immutability Multiplier
    if (isImm) {
      baseScore *= this.logic.multipliers.immutability;
    }
        
    // 3. Apply Blockchain Factors
    baseScore *= this.factors.exploit_maturity[maturity] || 1.0;
    baseScore *= this.factors.privileged_access[priv] || 1.0;
    baseScore *= this.factors.agent_autonomy[autonomy] || 1.0;
    
    // 4. Dynamic Scale to 0-10 (BVSS max possible)
    const maxImpact = 4;
    const maxLik = 3;
    const maxExpl = Math.max(...(Object.values(this.logic.bvss_metrics.exploitability_map) as number[]));
    const maxEcon = Math.max(...(Object.values(this.logic.bvss_metrics.economic_loss_map) as number[]));
    const maxBaseCalc = (maxImpact + maxLik + maxExpl + maxEcon) / 4;
    const maxScope = Math.max(...(Object.values(this.logic.bvss_metrics.scope_map) as number[]));
    const maxImm = this.logic.multipliers.immutability;
    const maxMat = Math.max(...(Object.values(this.factors.exploit_maturity) as number[]));
    const maxPriv = Math.max(...(Object.values(this.factors.privileged_access) as number[]));
    const maxAut = Math.max(...(Object.values(this.factors.agent_autonomy) as number[]));
    const maxPossible = maxBaseCalc * maxScope * maxImm * maxMat * maxPriv * maxAut;
    
    let scaledScore = (baseScore / maxPossible) * 10;
    scaledScore = Math.min(Math.max(Math.round(scaledScore * 10) / 10, 0), 10);
    
    // 5. Determine Severity
    let severity: 'Critical' | 'High' | 'Medium' | 'Low' = 'Low';
    let color = '🟢';
    for (const level of this.logic.severity_levels) {
      if (scaledScore >= level.threshold) {
        severity = level.label as 'Critical' | 'High' | 'Medium' | 'Low';
        color = level.color;
        break;
      }
    }
    
    // Set contextual remedy advice
    let remedy = '';
    if (scaledScore >= 9.0) {
      remedy = `⚠️ CRITICAL MITIGATION ADVICE: Vulnerability exhibits CRITICAL SEVERITY (Score: ${scaledScore.toFixed(1)}). Immutability multiplier is active, which guarantees permanent lock risk if contracts lack upgraded Proxy implementations. Cease active mint/redeem calls and apply emergency hot patches immediately.`;
    } else if (scaledScore >= 7.0) {
      remedy = `⚡ PRIORITY ACTION REQUIRED: Risk rating evaluated at HIGH SEVERITY (Score: ${scaledScore.toFixed(1)}). Remote exploitation over public networks requires no administrative privileges. Enforce robust formal verification checks using Move Prover immediately.`;
    } else if (scaledScore >= 4.0) {
      remedy = `📋 STANDARD MITIGATION RECOMMENDED: Overall risk rated at MEDIUM (Score: ${scaledScore.toFixed(1)}). Vulnerability commonly references dynamic field upgrades, clock manipulation, or oracle updates. Verify storage fund allocations.`;
    } else {
      remedy = `💡 AUDIT NOTE: Risk rating is LOW (Score: ${scaledScore.toFixed(1)}). Document in final audit reports and optimize code quality in upcoming upgrade sprints.`;
    }

    return {
      score: scaledScore,
      severity,
      color,
      remedy,
      breakdown: {
        impact: { label: impact, val: impactVal },
        likelihood: { label: lik, val: likelihoodVal },
        exploitability: { label: expl, val: explVal },
        economicLoss: { label: econ, val: econVal },
        scope: { label: scopeVal, mult: scopeMult },
        isImmutable: { active: isImm, mult: isImm ? this.logic.multipliers.immutability : 1.0 },
        exploitMaturity: { label: maturity, mult: this.factors.exploit_maturity[maturity] || 1.0 },
        privilegedAccess: { label: priv, mult: this.factors.privileged_access[priv] || 1.0 },
        agentAutonomy: { label: autonomy, mult: this.factors.agent_autonomy[autonomy] || 1.0 },
        baseScoreRaw: baseScore,
        maxPossible: maxPossible
      }
    };
  }
}

export function calculateBVSS(params: Partial<BVSSParams>, config: any): ScoreResult {
  const calculator = new BVSSCalculator(config);
  return calculator.calculate(params);
}
